pkg/products/rhssouser/reconciler.go

```diff
@@ -226,33 +319,241 @@ func (r *Reconciler) reconcileComponents(ctx context.Context, installation *inte
+		kcUser.Spec.User = user
+
+		return nil
+	})
+}
+
+func GetKeycloakUsers(ctx context.Context, serverClient k8sclient.Client, ns string) ([]keycloak.KeycloakAPIUser, error) {
+	var users keycloak.KeycloakUserList
+
+	listOptions := []k8sclient.ListOption{
+		k8sclient.MatchingLabels(getMasterLabels()),
+		k8sclient.InNamespace(ns),
+	}
+	err := serverClient.List(ctx, &users, listOptions...)
+	if err != nil {
+		return nil, err
+	}
+
+	mappedUsers := make([]keycloak.KeycloakAPIUser, len(users.Items))
+	for i, user := range users.Items {
+		mappedUsers[i] = user.Spec.User
+	}
+
+	return mappedUsers, nil
+}
+
+func getMasterLabels() map[string]string {
+	return map[string]string{
+		masterRealmLabelKey: masterRealmLabelValue,
+	}
+}
+
+func syncAdminUsersInMasterRealm(keycloakUsers []keycloak.KeycloakAPIUser, ctx context.Context, serverClient k8sclient.Client, ns string) ([]keycloak.KeycloakAPIUser, error) {
+
+	openshiftUsers := &usersv1.UserList{}
+	err := serverClient.List(ctx, openshiftUsers)
+	if err != nil {
+		return nil, err
+	}
+
+	openshiftAdminGroup := &usersv1.Group{}
+	err = serverClient.Get(ctx, k8sclient.ObjectKey{Name: "dedicated-admins"}, openshiftAdminGroup)
+	if err != nil {
+		return nil, err
+	}
+
+	added, deletedIndexes, deletedUsers := getUserDiff(keycloakUsers, openshiftUsers.Items, openshiftAdminGroup.Users)
+
+	for index := range deletedIndexes {
+		keycloakUsers = remove(index, keycloakUsers)
+	}
+
+	err = deleteKeycloakUsers(deletedUsers, ns, ctx, serverClient)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, osUser := range added {
+
+		if !isOpenshiftAdmin(osUser, openshiftAdminGroup) {
+			continue
+		}
+
+		email := osUser.Name
+		if !strings.Contains(email, "@") {
+			email = email + "@example.com"
+		}
+		keycloakUsers = append(keycloakUsers, keycloak.KeycloakAPIUser{
+			Enabled:       true,
+			UserName:      osUser.Name,
+			EmailVerified: true,
+			Email:         email,
+			FederatedIdentities: []keycloak.FederatedIdentity{
+				{
+					IdentityProvider: idpAlias,
+					UserID:           string(osUser.UID),
+					UserName:         osUser.Name,
+				},
+			},
+			RealmRoles: []string{"offline_access", "uma_authorization", "create-realm"},
+			ClientRoles: map[string][]string{
+				"account": {
+					"manage-account",
+					"view-profile",
+				},
+				"master-realm": {
+					"view-clients",
+					"view-realm",
+				},
+			},
+		})
+	}
+
+	return keycloakUsers, nil
+}
+
+func deleteKeycloakUsers(deletedUsers []keycloak.KeycloakAPIUser, ns string, ctx context.Context, serverClient k8sclient.Client) error {
+
+	var allUsers keycloak.KeycloakUserList
+
+	listOptions := []k8sclient.ListOption{
+		k8sclient.MatchingLabels(getMasterLabels()),
+		k8sclient.InNamespace(ns),
+	}
+
+	err := serverClient.List(ctx, &allUsers, listOptions...)
+	if err != nil {
+		return err
+	}
+
+	for _, delUser := range deletedUsers {
+
+		for _, user := range allUsers.Items {
+			if user.Spec.User.UserName == delUser.UserName {
```

Comment:
Maybe we should add a condition to check realm selector as well?
I know that right now, there are KeycloakUser(s) with master realm only, but if change this in the future, this could be a nasty bug.

Reply:
Or maybe we can match them with `.ID` ?

Reply:
We match by ID now where possible